src/dcap_provider.cpp

```diff
@@ -714,3 +721,112 @@ extern "C" sgx_plat_error_t sgx_ql_set_logging_function(
+        log(SGX_QL_LOG_INFO, "Fetching QE Identity from remote server: '%s'.", intel_cert_base_url.c_str());
+        curl->perform();
+
+        // content type
+        std::string content_type = curl->unescape(*curl->get_header(headers::CONTENT_TYPE));
+        log(SGX_QL_LOG_INFO, "CONTENT_TYPE:[%s]\n", content_type.c_str());
+
+        // issuer chain
+        const std::string* issuer_chain_raw_header = curl->get_header(headers::QE_ISSUER_CHAIN);
+        if (issuer_chain_raw_header == nullptr)
+        {
+            log(SGX_QL_LOG_ERROR,
+                "Header '%s' is missing.",
+                headers::QE_ISSUER_CHAIN);
+            return SGX_PLAT_ERROR_UNEXPECTED_SERVER_RESPONSE;
+        }
+        issuer_chain = curl->unescape(*issuer_chain_raw_header);
+        log(SGX_QL_LOG_INFO, "QE_ISSUER_CHAIN: [%s]\n", issuer_chain.c_str());
+
+        // requester id
+        std::string request_id = curl->unescape(*curl->get_header(headers::REQUEST_ID));
+        log(SGX_QL_LOG_INFO, "REQUEST_ID:[%s]\n", request_id.c_str());
+
+        // read body
+        identity_info = curl->get_body();
+        std::string qe_identity(curl->get_body().begin(), curl->get_body().end());
+        log(SGX_QL_LOG_INFO, "body:[%s]\n", qe_identity.c_str());
+
+        // Calculate total buffer size
+        total_buffer_size = safe_add(sizeof(sgx_qe_identity_info_t), identity_info.size());
+        total_buffer_size = safe_add(total_buffer_size, 1); // null terminator
+        total_buffer_size = safe_add(total_buffer_size, issuer_chain.size());
+
+        buffer = new char[total_buffer_size];
+        if (!buffer)
+            return SGX_PLAT_ERROR_OUT_OF_MEMORY;
+
+        memset(buffer, 0, total_buffer_size);
+
+        // fill in the qe info
+        p_qe_identity_info = reinterpret_cast<sgx_qe_identity_info_t*>(buffer);
+
+        // advance to the end of the sgx_qe_identity_info_t structure
+        buffer += sizeof(*p_qe_identity_info);
+
+        // qe_id_info
+        p_qe_identity_info->qe_id_info_size = static_cast<uint32_t>(identity_info.size());
+        p_qe_identity_info->qe_id_info = buffer;
+        memcpy(p_qe_identity_info->qe_id_info,
+                identity_info.data(),
+                identity_info.size());
+        buffer += identity_info.size() + 1; // skip null terminator
+
+        // set issuer_chain info
+        p_qe_identity_info->issuer_chain_size =  static_cast<uint32_t>(issuer_chain.size());
+        p_qe_identity_info->issuer_chain = buffer;
+        buffer += issuer_chain.size() + 1; // skip null terminator
+        memcpy(p_qe_identity_info->issuer_chain,
+                issuer_chain.data(),
+                issuer_chain.size());
+        *pp_qe_identity_info = p_qe_identity_info;
+    }
+    catch (std::bad_alloc&)
+    {
+        return SGX_PLAT_ERROR_OUT_OF_MEMORY;
+    }
+    catch (std::overflow_error& error)
+    {
+        log(SGX_QL_LOG_ERROR, "Overflow error. '%s'", error.what());
+        delete[] buffer;

```

Comment:
I'm pretty sure buffer will always be null at this point. However, it doesn't hurt to catch and clean up.

That said, if we're going to clean up in one handler, I think we should clean up in _all_ of them.

Reply:
An 2nd look at it, I agree that buffer will be null in the "catch std::overflow_error" block because all the safe_add operations happened above the memory allocation for buffer. I will remove the delete.